src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/FileWatchers/SolutionConfigFileWatcher.cs

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using NuGet.Configuration;
+
+namespace NuGet.PackageManagement.VisualStudio.Utility.FileWatchers
+{
+    internal class SolutionConfigFileWatcher : IFileWatcher
+    {
+        private bool _disposed;
+        private List<FileSystemWatcher> _watchers;
+
+        public SolutionConfigFileWatcher(string solutionDirectory)
+        {
+            string? current = solutionDirectory.EndsWith("\\", StringComparison.Ordinal)
+                ? Path.GetDirectoryName(solutionDirectory)
+                : solutionDirectory;
+
+            _watchers = new List<FileSystemWatcher>();
+            while (current != null)
+            {
+                if (Directory.Exists(current))
+                {
+                    var watcher = Create(current);
+                    _watchers.Add(watcher);
+                }
+
+                current = Path.GetDirectoryName(current);
+            }
+
+            FileSystemWatcher Create(string path)
+            {
+                var watcher = new FileSystemWatcher(path, Settings.DefaultSettingsFileName);
+                watcher.Created += OnFileSystemEvent;
+                watcher.Changed += OnFileSystemEvent;
+                watcher.Deleted += OnFileSystemEvent;
+                watcher.Renamed += OnFileSystemEvent;
+                watcher.EnableRaisingEvents = true;
+
+                return watcher;
+            }
+        }
+
+        public event EventHandler? FileChanged;
+
+        public void Dispose()
+        {
+            if (_disposed) { return; }
+
+            _disposed = true;
+
+            foreach (var watcher in _watchers)
+            {
+                watcher.EnableRaisingEvents = false;
+                watcher.Created -= OnFileSystemEvent;
+                watcher.Changed -= OnFileSystemEvent;
+                watcher.Deleted -= OnFileSystemEvent;
+                watcher.Renamed -= OnFileSystemEvent;
+                watcher.Dispose();
+            }
+
+            _watchers.Clear();
+
+            GC.SuppressFinalize(this);
+        }
+
+        private void OnFileSystemEvent(object sender, FileSystemEventArgs e)
+        {
+            FileChanged?.Invoke(this, EventArgs.Empty);
+        }
```

Comment:
I started with this as the implementation, but once I started writing tests, and thinking about all the cases to achieve (close to) 100% coverage, it added a lot of work for zero benefit, since the code that subscribes to IFileWatcher events wouldn't use the filename, even if it was in the event args. So, borrowing from the YAGNI principal, we just don't need it.

As we advance our diagnosability capabilities, I think it would be worthwhile having an ETW event, so that "report a problem" traces can tell us that config was reloaded. But that can be done within the two watcher implementations, it doesn't have to propagate the filename via event args.